### ISSUE TITLE
Vault API library should periodically force new connections in it's pool

### ISSUE THREAD

**Issue by chuckyz** [hashicorp/vault#15307]
**Is your feature request related to a problem? Please describe.**

When failing over via a DR it is very easy for an operator to forget the shutdown step.  When DNS is migrated to the newly failed over cluster there is a large chance of issues for a consumer to accidentally hold a connection open to the old cluster forever.

**Describe the solution you'd like**

Vault's API library should periodically reconnect it's HTTP connections.  

**Describe alternatives you've considered**

Placing a proxy in front of Vault to do this for us.

**Additional context**

We had a large outage, and having this feature would've saved us several hours worth of time.

**Comment by averche** [comment]
Hi @chuckyz! Thank you for bringing this to our attention! This is indeed a very interesting issue and it's hard to establish the exact root cause.

From what I understand, the vast majority of Vault client connections are short lived: most request wrappers read the entire response bodies til EOF and close them. The only exception I can think of is [`client.Sys().Monitor(ctx)`](https://github.com/hashicorp/vault/blob/main/api/sys_monitor.go#L14), which continues steaming server logs until its context has been explicitly cancelled by the caller. If the issue is indeed caused by `Monitor(ctx)`, the solution may be to periodically cancel the context and call the function again. This, however, risks dropping some of the monitored logs, so it may not be a great generic solution to implement in the library itself. Having said this, we are also working on a new OpenAPI-based Vault client library. With more freedom to make potentially breaking changes there, we could consider reworking how `Monitor(ctx)` works.

The other possibility that I can think of is that the issue is caused by the idle connections cached by `http.Transport`. From `net/http` [documentation](https://github.com/hashicorp/go-cleanhttp/blob/master/cleanhttp.go#L31):

> By default, Transport caches connections for future re-use. This may leave many open connections when accessing many hosts. This behavior can be managed using Transport’s CloseIdleConnections method and the MaxIdleConnsPerHost and DisableKeepAlives fields.

The default http client used by the Vault API library is `cleanhttp.DefaultPooledClient()`. It's a thin wrapper around `net/http`, which sets the [following](https://github.com/hashicorp/go-cleanhttp/blob/master/cleanhttp.go#L31):

```go
		MaxIdleConns:          100,
		IdleConnTimeout:       90 * time.Second,
		...
		MaxIdleConnsPerHost:   runtime.GOMAXPROCS(0) + 1,
```

Vault api library also uses [`go-retryablehttp`](https://github.com/hashicorp/go-retryablehttp), which does not close idle connections on successful requests.

If the idle connections are indeed the root cause, Vault client library allows overriding the http client and the retry behavior using the `Config` object. Using something like `cleanhttp.DefaultClient()` would disallow idle connections entirely. It might be worth trying it out just to see if that's the actual root cause. I would be hesitant to change the library's default HTTP client since it may adversely affect the performance. However, we could consider periodically calling `CloseIdleConnections()` on a separate go-routine if it would actually address the issue.

I'd be happy to investigate this further and work with you to debug the actual root cause! :raised_hands: 🤜 🤛

**Comment by chuckyz** [comment]
> However, we could consider periodically calling CloseIdleConnections() on a separate go-routine if it would actually address the issue.

This would be my ideal solution and I think would've helped us tremendously.  Something that is possible within C#'s SocketHttpHandler client is that you can even mark _active_ connections as 'as soon as this is idle, close it and force a new connection in the pool.'

Is there anything like that within `net/http` ?  That would be the absolute ideal case for us.

**Comment by averche** [comment]
Hi @chuckyz !

I tried to reproduce the issue using a fork of our official Go sample app, [hashicorp/hello-vault-go/sample-app](https://github.com/hashicorp/hello-vault-go/tree/main/sample-app). I [modified](https://github.com/averche/hello-vault-go/pull/1) the docker compose setup to use static IPs and a custom unbound DNS then ran a few tests to see if the connections are sticky. The full description of the tests I ran can be found in the updated [README.md](https://github.com/averche/hello-vault-go/blob/main/sample-app/README.md).

TLDR: from what I see, all connections established to Vault server by our test app through `vault/api` library enter `TIME_WAIT` state immediately after the request and are fully closed within a minute or so. Switching the DNS to point to a different vault server works as expected -- all new requests create new connections to the new IP:

```shell-session
$  docker exec -it sample-app-app-1 netstat -n | grep 8200
tcp        0      0 192.168.19.3:56208      192.168.19.4:8200       TIME_WAIT
tcp        0      0 192.168.19.3:60596      192.168.19.5:8200       TIME_WAIT   <- new connection
tcp        0      0 192.168.19.3:56200      192.168.19.4:8200       TIME_WAIT
tcp        0      0 192.168.19.3:56190      192.168.19.4:8200       TIME_WAIT
tcp        0      0 192.168.19.3:56212      192.168.19.4:8200       TIME_WAIT
```

I also tried adding a goroutine which calls `CloseIdleConnections()` every 5 seconds in https://github.com/averche/hello-vault-go/pull/1/commits/786c0df91fcbdcf27bc1430a26eb25f79ea6c3a7 . This did not seem to affect the connection behaviour. I'm guessing that's because the connections are not entering idle state. 

Please feel free to clone https://github.com/averche/hello-vault-go and follow the steps I described in the [README](https://github.com/averche/hello-vault-go/blob/main/sample-app/README.md). Let me know if I should be testing it differently, I'd be happy to modify the test further!

**Comment by hsimon-hashicorp** [comment]
Hi @chuckyz - since it's been a couple of months since we've heard from you, I'm going to go ahead and close this issue for now. If you want to keep working on it, please feel free to re-open it with the info requested by @averche and we'll be happy to keep investigating further. :)
Thanks for your understanding!
